R/call_standarise_formals.R

```diff
@@ -15,25 +15,135 @@ call_standardise_formals <- function(code, env = rlang::current_env(), include_d
   # out default formals. For primitives like mean, we're unable to distinguish
   # between mean() and mean.default()
   if (is_false(include_defaults) || is_infix(code) || is.primitive(fn)) {
-    return(rlang::call_standardise(code, env = env))
+    return(call_standardise_keep_partials(code, env = env))
   }
 
   fmls <- rlang::fn_fmls(fn)
   args_default <- fmls[!vapply(fmls, is.symbol, logical(1), USE.NAMES = FALSE)]
 
-  code_std <- rlang::call_standardise(code, env = env) # order and label existing params
+  # order and label existing params
+  code_std <- call_standardise_keep_partials(code, env = env)
 
-  # get arguments passed from user
+  # get named arguments passed from user
   args_user <- rlang::call_args(code_std)
+  args_user <- args_user[nzchar(names(args_user))]
 
-  args_default_missing <- setdiff(names(args_default), names(args_user))
+  args_default_missing <- names(args_default)[
+    !grepl(paste0("^", names(args_user), collapse = "|"), names(args_default))
+  ]
   if (length(args_default_missing) == 0) {
     return(code_std)
   }
 
   ## Add implicit default args to the call
-  rlang::call_standardise(
+  call_standardise_keep_partials(
     rlang::call_modify(code_std, !!!args_default[args_default_missing]),
     env = env
   )
 }
+
+call_standardise_keep_partials <- function(code, env = rlang::caller_env()) {
+  # If the function from the user code cannot be found, early exit because
+  #   standardization is impossible
+  fn <- tryCatch(call_fn(code, env), error = as.null)
+  if (is.null(fn)) return(code)
+
+  tryCatch(
+    rlang::call_match(code, fn),
+    error = function(e) {
+      # Check that error is caused by an ambiguous partial argument
+      # If not, return the code unaltered
```

Comment:
```suggestion
      # Check that error is caused by an ambiguous partial argument, which we
      # are forced to identify from its error message. The \Q...\E regex meta
      # characters mark the text inside them as literal. This way we can use the
      # literal, translated error message, but replace the `%d` with `\d+`.
```

Reply:
TIL about `\Q...\E` ! That's a pretty neat trick, but it took me a bit to work it out.